### PR TITLE
DataFrameDimension

### PR DESCRIPTION
**Pull request by jgrewe**

### CODE REVIEW

**Comment by achilleas-k** [comment]
LGTM.  Got a couple of issues for NIXPy out of this.

**Comment by achilleas-k** [comment]
@gicmo Will you have a look at this or should I go ahead and merge?
Not sure what's up with codecov and if we should try to restart it somehow until it's finished.

File: backend/hdf5/DimensionHDF5.cpp [hunk 1]
```diff
@@ -259,13 +269,153 @@ void SetDimensionHDF5::labels(const none_t t) {
...
 
+//--------------------------------------------------------------
+// Implementation of DataFrameDimensionHDF5
+//--------------------------------------------------------------
+DataFrameDimensionHDF5::DataFrameDimensionHDF5(const H5Group &group, ndsize_t index)
+    : DimensionHDF5(group, index)
+{
+    setType();
+    // Note: maybe having this is a bad idea, because a DataFrameDimension created
+    // this way will not have the block and file
```
Comment (gicmo) [inline backend/hdf5/DimensionHDF5.cpp:280]: Do we need it, if not remove it?

File: backend/hdf5/DimensionHDF5.cpp [hunk 2]
```diff
@@ -259,13 +269,153 @@ void SetDimensionHDF5::labels(const none_t t) {
...
+}
+
+DataFrameDimensionHDF5::DataFrameDimensionHDF5(const H5Group &group, ndsize_t index,
+                                               const std::shared_ptr<IFile> &file,
+                                               const std::shared_ptr<IBlock> &block, const DataFrame &frame)
+    :DataFrameDimensionHDF5(group, index, file, block)
+{
+    std::shared_ptr<IDataFrame> idf = block->getEntity<IDataFrame>(frame.id());
+    if (!idf)
+        throw std::runtime_error("ColumnDimensionHDF5 DataFrame not found in block!");
```
Comment (gicmo) [inline backend/hdf5/DimensionHDF5.cpp:298]: `ColumnDimensionHDF5` -> `DataFrameDimensionHDF5`

File: backend/hdf5/DimensionHDF5.cpp [hunk 3]
```diff
@@ -259,13 +269,153 @@ void SetDimensionHDF5::labels(const none_t t) {
...
+}
+
+int DataFrameDimensionHDF5::checkColumnIndex(int col_index) const {
+    int column_index = col_index;
+    if (col_index == -1) {
+        column_index = columnIndex();
+    }
+
+    if (column_index == -1) {
+        throw nix::OutOfBounds("DataFrameDimension: Error accessing column, no column index was given and no default is specified.");
```
Comment (gicmo) [inline backend/hdf5/DimensionHDF5.cpp:331]: Sometimes the error have `HDF5` suffix, sometimes not.

File: backend/hdf5/DimensionHDF5.cpp (hunk 1 above)
Comment (jgrewe) [reply backend/hdf5/DimensionHDF5.cpp:280]: maybe help me answering the question. RangeDimension and also the DataFrameDimension rely on file and block, but all Dimensions have this kind of constructor. 
I would rather remove the constructors.

File: backend/hdf5/DimensionHDF5.cpp (hunk 2 above)
Comment (jgrewe) [reply backend/hdf5/DimensionHDF5.cpp:298]: meh, good catch

File: backend/hdf5/DimensionHDF5.cpp (hunk 3 above)
Comment (jgrewe) [reply backend/hdf5/DimensionHDF5.cpp:331]: done, all get suffixes

File: backend/hdf5/DimensionHDF5.cpp (hunk 1 above)
Comment (jgrewe) [reply backend/hdf5/DimensionHDF5.cpp:280]: we can not remove the constructors, removed the note

File: backend/hdf5/DimensionHDF5.cpp [hunk 4]
```diff
@@ -259,13 +269,153 @@ void SetDimensionHDF5::labels(const none_t t) {
...
+
+
+std::string DataFrameDimensionHDF5::label(int col_index) const {
+    int column_index;
+    nix::DataFrame df = dataFrame();
+    try {
+        column_index = checkColumnIndex(col_index);
+    } catch (const nix::OutOfBounds& e) {
+        if (col_index == -1)
+            return df.name();
```
Comment (gicmo) [inline backend/hdf5/DimensionHDF5.cpp:365]: I think, if we just move that above `column_index = checkColumnIndex(col_index);` we can get rid of the `try-catch` block. I am not a huge fan of using exceptions for control flow in C++.

File: backend/hdf5/DimensionHDF5.cpp [hunk 5]
```diff
@@ -259,13 +269,153 @@ void SetDimensionHDF5::labels(const none_t t) {
...
+std::string DataFrameDimensionHDF5::unit(int col_index) const {
+    int column_index = checkColumnIndex(col_index);
+
+    nix::DataFrame df = dataFrame();
+    std::vector<Column> cols = df.columns();
+    return cols[column_index].unit;
+}
+
+
+std::string DataFrameDimensionHDF5::label(int col_index) const {
```
Comment (gicmo) [inline backend/hdf5/DimensionHDF5.cpp:358]: if `int` is used as a way to indicate that you might not actually want to select a column, maybe `boost::optional<unsigned>` would be better, so we don't have to mix types between signed and unsigned.

File: include/nix/DataArray.hpp [hunk 6]
```diff
@@ -381,6 +382,84 @@ class NIXAPI DataArray : public base::EntityWithSources<base::IDataArray>, publi
...
+     */
+    DataFrameDimension appendDataFrameDimension(const DataFrame &frame, const std::string &column_name) {
+        if (!frame) {
+            throw nix::UninitializedEntity();
+        }
+        unsigned index;
+        try {
+            std::vector<std::string> names = {column_name};
+            std::vector<unsigned> indices = frame.colIndex(names);
+            index = indices[0];
```
Comment (gicmo) [inline include/nix/DataArray.hpp:433]: There is `unsigned colIndex(const std::string &name) const` as a nonThere is `unsigned colIndex(const std::string &name) const` as a non-vector overload.

File: backend/hdf5/DimensionHDF5.cpp [hunk 7]
```diff
@@ -317,7 +470,7 @@ boost::optional<std::string> RangeDimensionHDF5::label() const {
 void RangeDimensionHDF5::label(const string &label) {
     H5Group g = redirectGroup();
     g.setAttr("label", label);
-    // NOTE: forceUpdatedAt() not possible since not reachable from here
+    // NOTE: forceUpdatedAt() not possible since not reacha?ble from here
```
Comment (gicmo) [inline backend/hdf5/DimensionHDF5.cpp:473]: `reacha?ble`

File: .travis.yml [hunk 8]
```diff
@@ -28,6 +28,10 @@ matrix:
       env: FS_BACKEND=OFF DOCKERFILE=ub1804
 
 addons:
+  apt:
+    update: true
+    packages:
+      - lcov
```
Comment (gicmo) [inline .travis.yml:34]: That is in theory not needed anymore, right @achilleas-k

File: backend/hdf5/DimensionHDF5.cpp [hunk 9]
```diff
@@ -259,13 +269,156 @@ void SetDimensionHDF5::labels(const none_t t) {
...
+    if (!col_index) {
+        column_index = columnIndex();
+    }
+
+    if (!column_index) {
+        throw nix::OutOfBounds("DataFrameDimensionHDF5: Error accessing column, no column index was given and no default is specified.");
+    }
+
+    nix::DataFrame df = dataFrame();
+    std::vector<Column> cols = df.columns();
```
Comment (gicmo) [inline backend/hdf5/DimensionHDF5.cpp:337]: Krass oversight in the `DataFrame` API that we have to fetch all the column information if all we want is the number. Filed as issue #808

File: backend/hdf5/DimensionHDF5.cpp [hunk 10]
```diff
@@ -259,13 +269,156 @@ void SetDimensionHDF5::labels(const none_t t) {
...
+    return cols[*column_index];
+}
+
+
+nix::DataType DataFrameDimensionHDF5::columnDataType(boost::optional<unsigned> col_index) const {
+    boost::optional<unsigned> column_index = checkColumnIndex(col_index);
+
+    nix::DataFrame df = dataFrame();
+    std::vector<Column> all_cols = df.columns();
```
Comment (gicmo) [inline backend/hdf5/DimensionHDF5.cpp:391]: Also lame that we have to fetch all of them if we only want one. Filed as issue #809

File: .travis.yml (hunk 8 above)
Comment (achilleas-k) [reply .travis.yml:34]: Correct.  Remove it.

Comment (gicmo) [reply .travis.yml:34]: No U!

**Pull request merged**